lib/rtti.h

```diff
@@ -0,0 +1,181 @@
+template <typename T>
+struct TypeIdResolver<T, std::void_t<decltype(T::static_typeId)>> {
+    static constexpr uint64_t resolve() noexcept {
+        constexpr TypeId id = T::static_typeId();
+        // Allow fall-back implementation if class does not want to explicitly
+        // specify a typeid
+        if constexpr (id == InvalidTypeId) return TypeIdResolver<T, int>::resolve();
+
+        return id;
+    }
+};
+
+}  // namespace detail
+
+struct Base;
+
+template <typename This, typename... Parents>
+struct TypeInfo {
+    using T = std::remove_const_t<std::remove_reference_t<This>>;
+
+    static_assert((... && std::is_base_of_v<Parents, This>),
+                  "One or more parents are not a base of this type.");
+
+    static_assert((... && std::is_base_of_v<Base, Parents>),
+                  "One or more parent hierarchies is not based on top of RTTI::Base.");
+
+    static_assert(std::is_base_of_v<Base, This>,
+                  "Invalid typeinfo requested, class is not based on top of RTTI::Base.");
+
+    static_assert(
+        std::is_same_v<T *, decltype(T::rttiEnabledMarker(nullptr))>,
+        "Invalid typeinfo requested, class does not declare typeinfo via DECLARE_TYPEINFO.");
+
+    [[nodiscard]] static constexpr TypeId id() noexcept {
+        return detail::TypeIdResolver<T>::resolve();
+    }
+
+    [[nodiscard]] static constexpr bool isA(TypeId typeId) noexcept {
+        return (id() == typeId) || (... || (Parents::TypeInfo::isA(typeId)));
+    }
+
+    template <typename T>
+    [[nodiscard]] static const void *dyn_cast(TypeId typeId, const T *ptr) noexcept {
+        if (id() == typeId) return static_cast<const This *>(ptr);
+
+        std::array<const void *, sizeof...(Parents)> ptrs = {
+            Parents::TypeInfo::dyn_cast(typeId, ptr)...};
+
+        auto it =
+            std::find_if(ptrs.begin(), ptrs.end(), [](const void *ptr) { return ptr != nullptr; });
+        return (it != ptrs.end()) ? *it : nullptr;
+    }
+};
+
+struct Base {
+    virtual ~Base() = default;
+
+    [[nodiscard]] virtual TypeId typeId() const noexcept = 0;
+    [[nodiscard]] virtual bool isA(TypeId typeId) const noexcept = 0;
+
+    template <typename T>
+    [[nodiscard]] bool is() const noexcept {
+        return isA(TypeInfo<T>::id());
+    }
+
+    template <typename T>
+    [[nodiscard]] T *to() noexcept {
+        return reinterpret_cast<T *>(const_cast<void *>(toImpl(TypeInfo<T>::id())));
+    }
+
+    template <typename T>
+    [[nodiscard]] const T *to() const noexcept {
+        return reinterpret_cast<const T *>(toImpl(TypeInfo<T>::id()));
+    }
+
+ protected:
+    [[nodiscard]] virtual const void *toImpl(TypeId typeId) const noexcept = 0;
+};
+
+}  // namespace RTTI
+
+#define DECLARE_TYPEINFO(T, ...)                                                  \
+ private:                                                                         \
+    static constexpr RTTI::TypeId static_typeId() { return RTTI::InvalidTypeId; } \
```

Comment:
Ah, it's an interesting question! Consider we are having the following class hierarchy:
```c++
class A {
  DECLARE_TYPEINFO_WITH_TYPEID(A, 42);
}; 

class B : public A {
  DECLARE_TYPEINFO(A, B);
};
```

So, `A` has explicit typeid and it is hash-based for `B`. Now we might have a problem, as `B` will have `static_typeId()` inherited from `A`. And therefore things will go very wrong. So we'd need to hide `static_typeId()` in `B`. And this is what this private declaration is doing.

We do have an examples of such usage in P4C, namely `NameMap<V>` and `NodeMap<K, V>`. Both derives from `IR::Node`, but has hash-generated typeids because their value types are not within `IR::Node` hierarchy, so we cannot do nested typeids.